packages/zoe/src/zoeService/purseStorage.js

```diff
@@ -0,0 +1,103 @@
+// @ts-check
+
+import { amountMath } from '@agoric/ertp';
+import { E } from '@agoric/eventual-send';
+import { makeWeakStore as makeNonVOWeakStore } from '@agoric/store';
+
+import { arrayToObj, objectMap } from '../objArrayConversion';
+
+export const makePurseStorage = () => {
+  /** @type {WeakStore<Brand, ERef<Purse>>} */
+  const brandToPurse = makeNonVOWeakStore('brand');
+
+  /**
+   * Create a purse for a new issuer
+   *
+   * @param {IssuerRecord} record
+   * @returns {void}
+   */
+  const createPurse = record => {
+    if (!brandToPurse.has(record.brand)) {
+      brandToPurse.init(record.brand, E(record.issuer).makeEmptyPurse());
+    }
+  };
+
+  /**
+   * Create a purse for a new, local issuer. Used only for ZCFMint issuers.
+   *
+   * @param {IssuerRecord} record
+   * @returns {Purse}
+   */
+  const createLocalPurse = record => {
+    const localPurse = record.issuer.makeEmptyPurse();
+    brandToPurse.init(record.brand, localPurse);
+    return localPurse;
+  };
+
+  /** @type {WithdrawPayments} */
+  const withdrawPayments = allocation => {
+    return harden(
+      objectMap(allocation, ([keyword, amount]) => {
+        const purse = brandToPurse.get(amount.brand);
+        return [keyword, E(purse).withdraw(amount)];
+      }),
+    );
+  };
+
+  /**
+   *
+   *  Only used internally. Actually deposit a payment or promise for payment.
+   *
+   * @param {ERef<Payment>} paymentP
+   * @param {Amount} amount
+   * @returns {Promise<Amount>}
+   */
+  const doDepositPayments = (paymentP, amount) => {
+    const purse = brandToPurse.get(amount.brand);
+    return E.when(paymentP, payment => E(purse).deposit(payment, amount));
+  };
+
+  // Proposal is cleaned, but payments are not
+
+  /**
+   * Deposits payments or promises for payments according to the
+   * `give` property of the proposal. Using the proposal, creates an
+   * initial allocation including the amount deposited for `give`
+   * keywords and an empty amount for `want` keywords.
```

Comment:
Something I've wondered about that I'd like to discuss someday: Could we make the `give` portion of a proposal optional? If omitted, could we derive it from depositing the payments?

Reply:
I think we get the following benefits from having a separate `give` section, which might be marginal: 1) we have a clear statement of intent that describes how much is being given declaratively, in somewhat human-readable fashion, that could be signed by a private key. 2) we stop users from accidentally sending more or different payments than they intend, if they checked the amounts in the proposal and those looked right. 3) If dapps or users are putting together a proposal, they are already going to have to describe what they are giving in terms of amounts that will later be withdrawn as payments, so it's not extra work in the usual user flow. The only exception to this is if the user immediately wants to use a payment they have lying around, but even in that case, they'll probably have checked the amount of that payment in some earlier step. 

The downsides: 1) we're replicating data that we know already and sending it to the chain unnecessarily. 2) there's a transaction cost to putting together the give part. 

So, I think there's no technical difficulty to omitting `give` entirely. But, the additional transaction costs for a entity (dapp or user) to put together `give` is minimal since we have to instruct the wallet how to withdraw the payments anyways. The additional data cost could end up being important, so we might want to revisit it later when we're optimizing that, but we also should reconsider the benefits at that point as well.

Reply:
These are compelling arguments. They do suggest that the error (below) for the extra payments is strictly better than either ignoring them or adding them as additional gives.

Reply:
Resolving this one, but leaving the error conversation below unresolved.